ocular/src/keyring.rs

```diff
@@ -0,0 +1,740 @@
+use bip32::PrivateKey;
+use bip39::{Mnemonic, Seed};
+use serde::{Deserialize, Serialize};
+use signatory::{
+    pkcs8::der::Document, pkcs8::EncodePrivateKey, pkcs8::LineEnding, pkcs8::PrivateKeyDocument,
+    FsKeyStore, KeyName,
+};
+use std::{env, fs, path::Path};
+
+use crate::error::KeyStoreError;
+
+// Constants
+// TODO: Move to independant constants file if reused elsewhere
+static COSMOS_BASE_DERIVATION_PATH: &str = "m/44'/118'/0'/0/0";
+static COSMOS_ADDRESS_PREFIX: &str = "cosmos";
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum KeyRingType {
+    File,
+    // TODO: OS (and other) Types
+}
+
+impl Default for KeyRingType {
+    fn default() -> Self {
+        KeyRingType::File
+    }
+}
+
+/// Basic keystore traits that all backends are expected to implement
+pub trait KeyStore {
+    /// Create new key store
+    fn create_key_store(&mut self) -> Result<(), KeyStoreError>;
+
+    /// Check if key store has been initialized.
+    fn key_store_created(&self) -> bool;
+
+    /// Check if key exists under specific name. Will return false if no key is found.
+    fn key_exists(&self, name: &str) -> Result<bool, KeyStoreError>;
+
+    /// Add a new key based off of name, password, and derivation path (defaults to cosmos). If override_if_exists is set to true, it will override any existing key with the same name.
+    fn add_key(
+        &self,
+        name: &str,
+        password: &str,
+        derivation_path: Option<&str>,
```

Comment:
I believe all Cosmos chains use the same coin type (118), otherwise every new app would have to apply for a derivation path. But we should get clarity on that from @jackzampolin or @zmanian